raiden/network/proxies/token_network.py

```diff
@@ -2175,73 +2271,171 @@ def settle(
-                    receipt_or_none = check_transaction_threw(self.client, transaction_hash)
+            transaction_hash = self.proxy.transact(
+                function_name="settleChannel",
+                startgas=gas_limit,
+                channel_identifier=channel_identifier,
+                **kwargs,
+            )
+            self.client.poll(transaction_hash)
+            receipt = check_transaction_threw(self.client, transaction_hash)
 
-            transaction_executed = gas_limit is not None
-            if not transaction_executed or receipt_or_none:
-                if transaction_executed:
-                    block = receipt_or_none["blockNumber"]
-                else:
-                    block = checking_block
+            if receipt:
+                failed_at_blockhash = encode_hex(receipt["blockHash"])
+                failed_at_blocknumber = receipt["blockNumber"]
 
                 self.proxy.jsonrpc_client.check_for_insufficient_eth(
                     transaction_name="settleChannel",
-                    transaction_executed=transaction_executed,
+                    transaction_executed=False,
                     required_gas=self.gas_measurements["TokenNetwork.settleChannel"],
-                    block_identifier=block,
+                    block_identifier=failed_at_blocknumber,
                 )
-                msg = self._check_channel_state_after_settle(
+
+                channel_onchain_detail = self._detail_channel(
                     participant1=self.node_address,
                     participant2=partner,
-                    block_identifier=block,
+                    block_identifier=failed_at_blockhash,
                     channel_identifier=channel_identifier,
                 )
-                raise RaidenUnrecoverableError(f"{error_prefix}. {msg}")
+                our_details = self._detail_participant(
+                    channel_identifier=channel_identifier,
+                    detail_for=self.node_address,
+                    partner=partner,
+                    block_identifier=failed_at_blockhash,
+                )
+                partner_details = self._detail_participant(
+                    channel_identifier=channel_identifier,
+                    detail_for=partner,
+                    partner=self.node_address,
+                    block_identifier=failed_at_blockhash,
+                )
+
+                our_balance_hash = hash_balance_data(
+                    transferred_amount=transferred_amount,
+                    locked_amount=locked_amount,
+                    locksroot=locksroot,
+                )
+                partner_balance_hash = hash_balance_data(
+                    transferred_amount=partner_transferred_amount,
+                    locked_amount=partner_locked_amount,
+                    locksroot=partner_locksroot,
+                )
+
+                if channel_identifier != channel_onchain_detail.channel_identifier:
```

Comment:
Hummm, this can be split up, could you please do:

- `if channel_onchain_detail.channel_identifier == 0 or channel_onchain_detail.channel_identifier > channel_identifier`.
  - With a 0 either the channel could have existed and it was settled already
  - With an channel id larger then the one used in the request, is the same as above, plus a new one was open.
  - In this branch then we can check if the channel ever existed [like this](https://github.com/raiden-network/raiden/blob/develop/raiden/network/proxies/payment_channel.py#L42-L55) and raise a `Recoverable` error.
- `elif channel_identifier != channel_onchain_detail.channel_identifier`. This is an unrecoverable error.

Reply:
Humm, the test for `on_chain_channel_id > channel_id` is sort of implied in the `channel_onchain_detail.state >= ChannelState.SETTLED`. Maybe these checks could be merged.